src/main/java/org/hbase/downstreamer/spark/JavaNetworkWordCountStoreInHBase.java

```diff
@@ -0,0 +1,269 @@
+          final File clientConfigs = new File("/etc/hbase/conf");
+          for (File siteConfig : clientConfigs.listFiles()) {
+            if (siteConfig.getName().endsWith(".xml")) {
+              LOG.debug("Adding config resource: {}", siteConfig);
+              conf.addResource(siteConfig.toURI().toURL());
+            }
+          }
+          UserGroupInformation ugi = null;
+          Connection connection = null;
+          synchronized(UserGroupInformation.class) {
+            LOG.debug("setting UGI conf");
+            UserGroupInformation.setConfiguration(conf);
+            LOG.debug("Ability to read keytab '{}' : {}", auth._2(), (new File(auth._2())).canRead());
+            LOG.debug("logging in via UGI and keytab.");
+            ugi = UserGroupInformation.loginUserFromKeytabAndReturnUGI(auth._1(), auth._2());
+            LOG.info("finished login attempt via UGI and keytab. security set? {}", ugi.isSecurityEnabled());
+          }
+          connection = ugi.doAs(new PrivilegedExceptionAction<Connection>() {
+            @Override
+            public Connection run() throws IOException {
+              return ConnectionFactory.createConnection(conf);
+            }
+          });
+          // Do something with the connection now, to ensure we have valid credentials.
+          final Admin admin = connection.getAdmin();
+          try {
+            final ClusterStatus status = admin.getClusterStatus();
+            LOG.info("connection successful: {}", status);
+            final Tuple2<UserGroupInformation, Connection> proposed = new Tuple2<UserGroupInformation, Connection>(ugi, connection);
+            final Tuple2<UserGroupInformation, Connection> prior = connections.putIfAbsent(auth, proposed);
+            if (prior == null) {
+              // our proposed connection is valid.
+              result = proposed;
+            } else {
+              // parallel instantiation beat us to completion
+              LOG.warn("Discarding extra connection. No need to be concerned, unless this message happens dozens of times.");
+              result = prior;
+              connection.close();
+            }
+          } finally {
+            admin.close();
+          }
+        } catch (IOException exception) {
+          LOG.error("Failed to connect to hbase. rethrowing; application should fail.", exception);
+          throw exception;
+        }
+      }
+      return result;
+    }
+
+    @Override
+    public void call(Iterator<Tuple2<String, Integer>> iterator) throws IOException, InterruptedException {
+      try {
+        Tuple2<UserGroupInformation, Connection> hbase = ensureConnection(auth);
+        // This presumes we can complete our Put calculation before the TGT expires.
+        // At worst after this check we will have 20% of the ticket window left; for 24hr tickets that means
+        // just under 5 hours.
+        hbase._1().checkTGTAndReloginFromKeytab();
```

Comment:
This shouldn't be needed, because when we're logged in via keytab the RPC handler within the HBase Connection should attempt a relogin should things fail. However, in HBase's internal Canary we still make a renewal thread to handle the same thing we're doing here. So we'll need to test removing it. And then update the HBase docs to make clear what downstream apps ought to be doing. :)